File: data_structure(20-1)/7.queue/7주차_큐.py
#함수 정의 및 선언

def UserSetting():#사용자에게 프로그램 사용법을 표시해주는 함수
    print("사용자는 다음의 명령어를 사용할 수 있습니다.")
    print()
    print("I(대문자 i) : enqueue 명령(+enqueue할 요소)")
    print("L : 큐 전체 출력")
    print("P : head 위치의 요소 peek(보기)")
    print("- : head 위치의 요소 하나만 dequeue 명령(dequeue 후 자동적으로 앞으로 정렬됨)")
    print("-5 : head 위치의 요소부터 5개까지 dequeue(5가 아니어도 됨)")
    print("F : 큐가 꽉 찼는지 여부")
    print("# : 큐 요소 개수 확인")
    print("H : head의 현재 위치 출력")
    print("T : tail의 현재 위치 출력")
    print("E : 빈 큐 여부")
    print("?a : 큐에 요소가 있는지 여부(a가 아니어도 됨)")
    print("=a : tail 위치에 있는 요소를 사용자가 입력한 요소로 변경(a가 아니어도 됨)")
    print("C : 큐 비우기(요소들 보여주면서 비워짐)")
    print("Q : 프로그램 종료")
    print()
    print("밑의 3개는 자신만의 기능")
    print("S : 앞으로 정렬(앞에 빈 자리가 없도록 요소들을 이동)")
    print("Z : tail 뒤에 남은 자리가 얼마인지 출력")
    print("R : 큐의 순서를반대로 해서 출력(큐의 상태는 그대로)")
    print()
    print("PrintAll 함수와 is_member 함수에서 head를 원래대로 하는 연산을 수정했습니다.")
    print()

def Setposition():#첫 현재 위치 설정
    return -1

def DataCount(head, tail):#큐 요소 개수
    return tail-head

def Shift(array,head,tail):#앞으로 정렬
    if head==-1:#이미 앞으로 정렬된 상태이면 그대로 반환
        return
    
    else:
        frontlen=head+1#head 앞에 남은 자리의 개수
        i=0
        while head!=tail:#요소들을 모두 앞으로 당김
            head+=1
            array[i]=array[head]
            del array[head]
            i+=1

        head=Setposition()#head와 tail 위치 재설정
        tail-=frontlen

def Enqueue(array,tail):#enqueue 함수
    temp=input("abcd의 형식으로 삽입할 요소를 입력하세요: ")
    enlen=len(temp)#사용자가 넣고 싶은 요소의 개수(ex.abcd면 4개)
    remainlen=Max_Size-tail-2#큐에 남은 자리의 총 개수

    if (enlen>remainlen):#enlen보다 remainlen이 부족할 때
        print("큐에 남은 자리가 부족합니다. 남은 자리는 %d개입니다."%remainlen)
            
    else:#큐에 자리가 넉넉할 때
        for i in range(enlen):
            array.append(temp[i])

        tail+=enlen

    return tail

def is_empty(head,tail):#큐가 비었는지
    if head==tail:
        return True
    else:
        return False

def PrintAll(array,head,tail):#현재 배열 상황을 출력해주는 함수

    temp=[]
    
    if is_empty(head,tail)==True:#큐가 비었으면
        print("출력할 것이 없음")
        
    else:#큐에 저장된 것이 있으면
        head+=1
        while head!=(tail+1):
            print(array[head],end='')#head요소를 dequeue하고
            temp.append(array[head])#temp에 저장
            del array[head]
            tail-=1
        print()

        #이후 temp에 넣어뒀던 것들을 다시 큐에 넣음
        #enqueue의 과정과 동일

        enlen=len(temp)#사용자가 넣고 싶은 요소의 개수(ex.abcd면 4개)
        remainlen=Max_Size-tail-2#큐에 남은 자리의 총 개수

        if (enlen>remainlen):#enlen보다 remainlen이 부족할 때
            print("큐에 남은 자리가 부족합니다. 남은 자리는 %d개입니다."%remainlen)
            
        else:#큐에 자리가 넉넉할 때
            for i in range(enlen):
                array.append(temp[i])

            tail+=enlen

    return tail

def Peek(array,head,tail):#peek 함수
    if is_empty(head,tail)==True:#큐가 비었으면
        print("출력할 것이 없음")
    else:
        savehead=head
        head+=1
        print(array[head])
        head=savehead

def Dequeue(array,order,head,tail):#dequeue 함수(c와는 달리, python은 요소들이 계속해서 앞으로 정렬되는 것을 이용)
    if len(order)!=1:#-숫자 형태로 입력받은 경우
        if order[1].isdigit()==True:#-이후 사용자가 정수를 입력한 경우
            num=int(order[1])
            if num>DataCount(head,tail):
                print("입력한 숫자가 큐에 있는 요소 개수보다 많습니다. 다시 입력하세요",end='')
            else:
                for i in range(num):
                    savehead=head
                    head+=1
                    print(array[head],end='')#peek하고
                    del array[head]#삭제
                    head=savehead
                    tail-=1
                    
    else:#-만 입력받은 경우
        if is_empty(head,tail)==True:#큐가 비었으면
            print("dequeue할 것이 없음",end='')
        else:#큐에 저장된 것이 있는 경우
            savehead=head
            head+=1
            print(array[head],end='')
            del array[head]
            head=savehead
            tail-=1

    return tail

def is_full(tail):#큐가 가득찼는지
    if tail==Max_Size-2:
        return True
    else:
        return False

def Head(head):#head
    return head

def Tail(tail):#tail
    return tail

def is_member(array,order,head,tail):#큐에 요소가 있는지 여부
    temp=[]
    
    if is_empty(head,tail)==1:#큐가 비어있으면
        print("False")
        
    else:
        head+=1
        while head!=(tail+1):
            if array[head]==order[1]:#요소를 찾으면
                print("True")
            temp.append(array[head])#temp에 저장
            del array[head]
            tail-=1

        #이후 temp에 넣어뒀던 것들을 다시 큐에 넣음
        #enqueue의 과정과 동일

        enlen=len(temp)#사용자가 넣고 싶은 요소의 개수(ex.abcd면 4개)
        remainlen=Max_Size-tail-2#큐에 남은 자리의 총 개수

        if (enlen>remainlen):#enlen보다 remainlen이 부족할 때
            print("큐에 남은 자리가 부족합니다. 남은 자리는 %d개입니다."%remainlen)
            
        else:#큐에 자리가 넉넉할 때
            for i in range(enlen):
                array.append(temp[i])

            tail+=enlen

    return tail

def Replace(array,order,head,tail):#변경
    if is_empty(head,tail)!=1:
        array[tail]=order[1]

def Clear(array,order,head,tail):#비우기
    while head!=tail:
        tail=Dequeue(array,order,head,tail)#처음부터 끝까지 dequeue

    print()
    return tail

def Rearlen(tail):#tail 뒤에 남은 자리가 얼마인지 출력
    return Max_Size-tail-2

def Reverse(array,head,tail):#큐의 순서를 반대로 해서 출력
    if is_empty(head,tail)==1:
        print("뒤집을 것이 없음")
    else:
        temp=[]
        savehead=head
        while head!=tail:
            head+=1
            temp.append(array[head])#큐의 요소들을 temp에 저장
        head=savehead

        for i in range(len(temp)//2):#temp를 거꾸로
            temp[i],temp[len(temp)-1-i]=temp[len(temp)-1-i],temp[i]

        for i in range(len(temp)):
            print(temp[i],end='')
        print()
            
#메인 함수
array=[]#배열을 하나 생성함

#첫 head와 tail 설정
head=Setposition()
tail=Setposition()

UserSetting()#프로그램 사용법 출력

Max_Size=100#임의로 Max_Size 설정

while(True):#Q명령 시 종료하고, 그 이외에는 프로그램이 계속 실행될 수 있도록 while문 사용
    order=input("명령어를 입력하세요: ")#사용자에게 명령어를 받음
    
    if (order=='Q'):#프로그램을 종료하는 명령어
        print("프로그램을 종료합니다.")
        break

    if (order=='I'):#enqueue
        tail=Enqueue(array,tail)
        
    if (order=='L'):#현재 배열 출력
        tail=PrintAll(array,head,tail)

    if (order=='P'):#peek
        Peek(array,head,tail)

    if (order[0]=='-'):#dequeue
        tail=Dequeue(array,order,head,tail)
        print()

    if (order=='F'):#is_full
        print(is_full(tail))

    if (order=='#'):#개수 확인
        print(DataCount(head,tail))

    if (order=='H'):#head 현재 위치
        print(Head(head))

    if (order=='T'):#tail 현재 위치
        print(Tail(tail))

    if (order=='E'):#빈 큐 여부
        print(is_empty(head,tail))

    if (order[0]=='?'):#is_member
        tail=is_member(array,order,head,tail)

    if (order[0]=='='):#replace
        Replace(array,order,head,tail)

    if (order=='C'):#clear
        tail=Clear(array,order,head,tail)

    if (order=='S'):#shift
        Shift(array,head,tail)

    if (order=='Z'):#rearlen
        print(Rearlen(tail))

    if (order=='R'):#reverse
        Reverse(array,head,tail)

File: data_structure/7.queue/7주차_큐.c
#pragma warning(disable: 4996)
#include<stdio.h>
#include<string.h>//문자열 길이, 복사 위한 헤더
#include<stdbool.h>//true, false

#define Max_Size 30 // 큐의 크기를 30으로 static하게 설정(마지막은 '\0'가 있어야 하므로 실질적 크기는 29)

void UserSetting()
{
	printf("사용자는 다음의 명령어를 사용할 수 있습니다.\n\n");
	printf("+abcd : enqueue 명령(enqueue할 요소는 마음대로)\n");
	printf("L : 큐 전체 출력\n");
	printf("P : head 위치의 요소 peek(보기)\n");
	printf("- : head 위치의 요소 하나만 dequeue 명령\n");
	printf("-5 : head 위치의 요소부터 5개까지 dequeue(5가 아니어도 됨)\n");
	printf("F : 큐가 꽉 찼는지 여부\n");
	printf("# : 큐 요소 개수 확인\n");
	printf("H : head의 현재 위치 출력\n");
	printf("T : tail의 현재 위치 출력\n");
	printf("E : 빈 큐 여부\n");
	printf("?a : 큐에 요소가 있는지 여부(a가 아니어도 됨)\n");
	printf("=a : tail 위치에 있는 요소를 사용자가 입력한 요소로 변경(a가 아니어도 됨)\n");
	printf("C : 큐 비우기(요소들 보여주면서 비워짐)\n");
	printf("Q : 프로그램 종료\n\n");
	printf("밑의 3개는 자신만의 기능\n");
	printf("S : 앞으로 정렬(앞에 빈 자리가 없도록 요소들을 이동)\n");
	printf("Z : tail 뒤에 남은 자리가 얼마인지 출력\n");
	printf("R : 큐의 순서를 반대로 해서 출력(큐의 상태는 그대로)\n\n");
	printf("PrintAll 함수와 is_member 함수에서 head를 원래대로 하는 연산을 수정했습니다.\n\n");

}
int Setposition()//head의 처음 위치 설정
{
	return -1;//큐에 아무것도 들어있지 않을 때, head와 tail은 -1
}

typedef struct _position
{
	int head;
	int tail;

}Position;

void Positioninit(Position* p)
{
	p->head = Setposition();
	p->tail = Setposition();
}


int DataCount(Position* p)//큐 요소 개수
{
	return p->tail - p->head;
}

void Shift(char array[], Position* p)
{
	if (p->head == -1)//이미 앞으로 정렬된 상태이면 그대로 반환
		return;

	int frontlen = p->head + 1;//head 앞에 남은 자리의 개수
	int i = 0;

	while (p->head != p->tail) {//요소들을 모두 앞으로 당김
		p->head++;
		array[i] = array[p->head];
		array[p->head] = '\0';
		i++;
	}

	p->head = Setposition();//head와 tail 위치 재설정
	p->tail -= frontlen;
}

void Enqueue(char array[], char order[], Position* p)//enqueue함수
{
	char temp_array[Max_Size];
	int enlen, rearlen, frontlen, remainlen;

	int i;

	for (i = 0; i < strlen(order) - 1; i++) {//+abcd를 temp_array에 abcd로 저장
		temp_array[i] = order[i + 1];
	}
	temp_array[strlen(order) - 1] = '\0';

	enlen = strlen(temp_array);//사용자가 넣고 싶은 요소의 개수(ex.abcd면 4개)
	rearlen = Max_Size - p->tail - 2;//tail 뒤에 남은 자리의 개수
	frontlen = p->head + 1;//head 앞에 남은 자리의 개수
	remainlen = frontlen + rearlen;//큐에 남은 자리의 총 개수

	if (enlen > rearlen) {//enlen보다 tail 뒤의 자리가 부족할 때(ex. 4개 넣고 싶은데 tail 뒤에 3자리밖에 없을 때)
		if (enlen > remainlen)//enlen보다 remainlen이 부족할 때
			printf("큐에 남은 자리가 부족합니다. 남은 자리는 %d개입니다.\n", remainlen);
		else {//남은 자리의 총 개수가 모자르지 않을 때
			Shift(array, p);//앞에 빈자리가 없도록 큐 정렬

			for (i = 0; i < strlen(temp_array); i++) {//tail 뒤에 입력받은 것들 넣음
				array[p->tail + 1 + i] = temp_array[i];
			}

			p->tail += strlen(temp_array);
		}
	}

	else {//큐에 자리가 넉넉할 때
		for (i = 0; i < strlen(temp_array); i++) {//tail 뒤에 입력받은 것들 넣음
			array[p->tail + 1 + i] = temp_array[i];
		}

		p->tail += strlen(temp_array);
	}
}

void PrintAll(char array[], Position* p)//큐 전체 출력 함수
{
	char temp_array[Max_Size] = { '\0', };
	int i = 0;
	int enlen, rearlen, frontlen, remainlen;

	if (p->head == p->tail)//큐가 비었으면
		printf("출력할 것이 없음\n");

	else {//큐에 저장된 것이 있는 경우
		while (p->head != p->tail)
		{
			p->head++;
			printf("%c", array[p->head]);//head요소를 peek하고
			temp_array[i] = array[p->head];//temp_array에 저장
			i++;
		}
		printf("\n");

		//이후 temp에 넣어뒀던 것들을 다시 큐에 넣음
		//enqueue의 과정과 동일

		enlen = strlen(temp_array);//넣고 싶은 요소의 개수(ex.abcd면 4개)
		rearlen = Max_Size - p->tail - 2;//tail 뒤에 남은 자리의 개수
		frontlen = p->head + 1;//head 앞에 남은 자리의 개수
		remainlen = frontlen + rearlen;//큐에 남은 자리의 총 개수

		if (enlen > rearlen) {//enlen보다 tail 뒤의 자리가 부족할 때(ex. 4개 넣고 싶은데 tail 뒤에 3자리밖에 없을 때)
			if (enlen > remainlen)//enlen보다 remainlen이 부족할 때
				printf("큐에 남은 자리가 부족합니다. 남은 자리는 %d개입니다.\n", remainlen);
			else {//남은 자리의 총 개수가 모자르지 않을 때
				Shift(array, p);//앞에 빈자리가 없도록 큐 정렬

				for (i = 0; i < strlen(temp_array); i++) {//tail 뒤에 입력받은 것들 넣음
					array[p->tail + 1 + i] = temp_array[i];
				}

				p->tail += strlen(temp_array);
			}
		}

		else {//큐에 자리가 넉넉할 때
			for (i = 0; i < strlen(temp_array); i++) {//tail 뒤에 입력받은 것들 넣음
				array[p->tail + 1 + i] = temp_array[i];
			}

			p->tail += strlen(temp_array);
		}
	}
}

void Peek(char array[], Position* p)//head에 있는 것을 출력하는 peek함수
{
	if (p->head == p->tail)//큐가 비었으면
		printf("peek할 것이 없음\n");
	else//큐에 저장된 것이 있는 경우
	{
		int savehead = p->head;
		p->head++;
		printf("%c\n", array[p->head]);
		p->head = savehead;
	}
}

void Dequeue(char array[], char order[], Position* p)//dequeue함수(한번에 하나씩만 꺼낼 수도, 숫자를 이용해 여러 개 꺼낼 수도 있음)
{
	char s1[30];
	int num;

	if (order[1] != '\0') {//-숫자 형태로 입력받은 경우
		for (int i = 0; i < strlen(order) - 1; i++) {//예를 들어 사용자가 -12를 입력했으면, 12를 s1에 문자 형태로 저장
			s1[i] = order[i + 1];
		}
		s1[strlen(order) - 1] = '\0';

		num = atoi(s1);//12를 숫자로 변환

		if (num != 0) {//-이후 사용자가 정수를 입력한 경우
			if (num > DataCount(p))
				printf("입력한 숫자가 큐에 있는 요소 개수보다 많습니다. 다시 입력하세요");
			else {
				for (int i = 0; i < num; i++) {
					p->head++;
					printf("%c", array[p->head]);//peek하고
					array[p->head] = '\0';//삭제
				}
				printf(" 삭제");
			}
		}
	}

	else {//-만 입력받은 경우
		if (p->head == p->tail)//큐가 비었으면
			printf("dequeue할 것이 없음\n");

		else {//큐에 저장된 것이 있는 경우
			p->head++;
			printf("%c", array[p->head]);//peek하고
			array[p->head] = '\0';//삭제
		}
	}
}

bool is_full(Position* p)//큐 가득 찼는지 여부
{
	if (p->tail == Max_Size - 2)
		return true;
	else
		return false;
}


int Head(Position* p)
{
	return p->head;
}

int Tail(Position* p)
{
	return p->tail;
}

bool is_empty(Position* p)
{
	if (p->head == p->tail)
		return true;
	else
		return false;
}

bool is_member(char array[], char order[], Position* p)//큐에 요소가 있는지 여부 함수
{
	char temp_array[Max_Size] = { '\0', };
	int i = 0;
	int enlen, rearlen, frontlen, remainlen;
	bool exist = false;//요소 존재 여부 확인하는 변수

	if (is_empty == 1)//큐에 저장된 것이 없는 경우
		printf("False\n");

	else {//큐에 저장된 것이 있는 경우
		while (p->head != p->tail)
		{
			p->head++;
			if (array[p->head] == order[1])//큐에 요소가 존재하면
				exist = true;
			temp_array[i] = array[p->head];//temp_array에 저장
			i++;
		}

		//head를 원래대로 하는 것이 아닌, enqueue의 과정이 이루어지도록 함

		enlen = strlen(temp_array);//넣고 싶은 요소의 개수(ex.abcd면 4개)
		rearlen = Max_Size - p->tail - 2;//tail 뒤에 남은 자리의 개수
		frontlen = p->head + 1;//head 앞에 남은 자리의 개수
		remainlen = frontlen + rearlen;//큐에 남은 자리의 총 개수

		if (enlen > rearlen) {//enlen보다 tail 뒤의 자리가 부족할 때(ex. 4개 넣고 싶은데 tail 뒤에 3자리밖에 없을 때)
			if (enlen > remainlen)//enlen보다 remainlen이 부족할 때
				printf("큐에 남은 자리가 부족합니다. 남은 자리는 %d개입니다.\n", remainlen);
			else {//남은 자리의 총 개수가 모자르지 않을 때
				Shift(array, p);//앞에 빈자리가 없도록 큐 정렬

				for (i = 0; i < strlen(temp_array); i++) {//tail 뒤에 입력받은 것들 넣음
					array[p->tail + 1 + i] = temp_array[i];
				}

				p->tail += strlen(temp_array);
			}
		}

		else {//큐에 자리가 넉넉할 때
			for (i = 0; i < strlen(temp_array); i++) {//tail 뒤에 입력받은 것들 넣음
				array[p->tail + 1 + i] = temp_array[i];
			}

			p->tail += strlen(temp_array);
		}
	}

	return exist;
}

void Replace(char order[], char array[], Position* p)//tail 요소 변경 함수
{
	if (is_empty(p) != 1)
		array[p->tail] = order[1];
}

void Clear(char array[], char order[], Position* p)//큐 비우기 함수
{
	while (p->head != p->tail) {
		Dequeue(array, order, p);
	}
	printf("\n");

	Positioninit(p);
}

int Z(Position* p)//tail 뒤에 남은 자리의 개수
{
	int rearlen = Max_Size - p->tail - 2;
	return rearlen;
}

void Reverse(char array[], Position* p)//뒤집고 출력 함수
{
	char temp[Max_Size] = { '\0' };
	int savehead = p->head;//원래의 head위치 저장
	int i = 0;

	if (is_empty(p) == 1)//큐가 비었으면
		printf("뒤집을 것이 없음\n");

	else {//큐에 저장된 것이 있는 경우
		while (p->head != p->tail)
		{
			p->head++;
			temp[i] = array[p->head];
			i++;
		}
		p->head = savehead;//head를 원래대로

		int cnt = strlen(temp);
		char tmp;//단어 하나를 임시로 저장할 변수

		for (i = 0; i < cnt / 2; i++) {// 교환이 중복되지 않기 위해 cnt/2까지 반복함
			tmp = temp[i];
			temp[i] = temp[cnt - 1 - i];
			temp[cnt - 1 - i] = tmp;
		}// 0번과 끝번, 1번과 끝-1번...순으로 교환함

		printf("%s\n", temp);
	}
}

int main() {
	char array[Max_Size] = { '\0', };
	char order[30];
	Position position;

	Positioninit(&position);

	UserSetting();

	while (1) {

		printf("order : ");
		scanf("%s", order);

		if (order[0] == 'Q') {
			printf("프로그램을 종료합니다.\n");
			break;
		}//escape code

		if (order[0] == '+') {
			Enqueue(array, order, &position);
		}

		if (order[0] == 'L') {
			PrintAll(array, &position);
		}

		if (order[0] == 'P') {
			Peek(array, &position);
		}

		if (order[0] == '-') {
			Dequeue(array, order, &position);
			printf("\n");
		}

		if (order[0] == 'F') {
			if (is_full(&position) == 1)
				printf("True\n");
			else
				printf("False\n");
		}

		if (order[0] == '#') {
			printf("%d\n", DataCount(&position));
		}

		if (order[0] == 'H') {
			printf("%d\n", Head(&position));
		}

		if (order[0] == 'T') {
			printf("%d\n", Tail(&position));
		}

		if (order[0] == 'E') {
			if (is_empty(&position) == 1)
				printf("True\n");
			else
				printf("False\n");
		}

		if (order[0] == '?') {
			if (is_member(array, order, &position) == 1)
				printf("True\n");
			else
				printf("False\n");
		}

		if (order[0] == '=') {
			Replace(order, array, &position);
		}

		if (order[0] == 'C') {
			Clear(array, order, &position);
		}

		if (order[0] == 'S') {
			Shift(array, &position);
		}

		if (order[0] == 'Z') {
			printf("%d\n", Z(&position));
		}

		if (order[0] == 'R') {
			Reverse(array, &position);
		}
	}

	system("pause");
	return 0;
}